### PR TITLE
Add feature for fetching authorised SSH keys from S3

### PR DESCRIPTION
**Pull request by niklasvincent**
We need the ability to replace `~/.ssh/authorized_keys`, instead of relying on a shared SSH key.

This PR adds an optional feature that can be called as part of server bootstrapping to populate `~/.ssh/authrorized_keys` for a specific user, e.g:

`./ssh-keys/fetch-authorized-keys.sh -k s3://mybucket/authorized_keys -u ubuntu`

At first I thought about just appending the keys, but I settled on overwriting it. The advantage of overwriting is the ability to revoke keys by removing them from the S3 file or add new keys on demand.

### CODE REVIEW

**Comment by cb372** [comment]
That's an impressively engineered script, given that we (CAPI team) currently do this with a single call to `wget` :)

Could you name the script `install.sh` for consistency with other features please. Apart from that, :+1:

**Comment by mchv** [comment]
As @cb372 mentioned the code look quite complicated for the feature. In CAPI I think that we achieve the same using:

```
wget -NP /home/ubuntu/.ssh https://s3-eu-west-1.amazonaws.com/content-api-dist/authorized_keys
```

**Comment by mchv** [comment]
@nlindblad another feature you might find useful is to schedule the override every hour:

``` bash
 cat > /etc/cron.d/sync-ssh-keys << EOF
 # Updated ssh keys every hour
 23 * * * * root wget -NP /home/ec2-user/.ssh https://s3-eu-west-1.amazonaws.com/discussion- dist/authorized_keys
 EOF
```

**Comment by niklasvincent** [comment]
I agree that wget would work for an existing user, with an existing `~/.ssh/authorized_keys`, with the correct permissions (which should be the default for a user that AWS adds a key pair to as part of deployment).

The motivation behind this script is that it will work for newly created users that do not have an `~/.ssh` folder yet. It will also ensure the right file and directory permissions, which is a common cause of annoying OpenSSH problems.

I've updated the script to do a `chown` as well as ensuring the user has an `~/.ssh` directory.

The crontab entry is a good idea. Thanks @mchv!

**Comment by mchv** [comment]
@nlindblad Thanks for taking the time to reply. I better understand the issues you mention and and the motivation.

**Comment by niklasvincent** [comment]
@mchv No problem. I think a department wide discussion about SSH key provisioning might be useful, because it seems like every team does it slightly different and there's probably a lot to be learnt from each other (like the crontab entry, we currently don't have anything like that).

@mchv @cb372 Are you fine with me merging this in?

File: packer/resources/features/ssh-keys/install.sh [hunk 1]
```diff
@@ -0,0 +1,146 @@
...
+if [ -z "${S3_KEY_URL}" ]; then
+  echo "Must specify an S3 URL"
+  exit 1
+fi
+
+if [ -z "${SSH_USER}" ]; then
+  SSH_USER="ubuntu"
+fi
+
+TEMPORARY_SSH_KEY_FILE="/tmp/authorized_keys.$$"
```
Comment (cb372) [inline packer/resources/features/ssh-keys/install.sh:48]: What's the reason for the temp file? Why not download directly to `~/.ssh/authorized_keys`?

Comment (niklasvincent) [inline packer/resources/features/ssh-keys/install.sh:48]: To handle network problems/temporary hiccups with S3 reasonably. Overwriting directly might result in a partial download or no data at all in `authorized_keys`.

**Comment by cb372** [comment]
:+1: feel free to merge

**Pull request merged**